tools/pythonpkg/pyduckdb/spark/sql/types.py

```diff
@@ -1,9 +1,1221 @@
+        if self._needSerializeAnyField:
+            # Only calling fromInternal function for fields that need conversion
+            values = [
+                f.fromInternal(v) if c else v
+                for f, v, c in zip(self.fields, obj, self._needConversion)
+            ]
+        else:
+            values = obj
+        return _create_row(self.names, values)
+
+class UnionType(DataType):
+	def __init__(self):
+		raise ContributionsAcceptedError
+
+class UserDefinedType(DataType):
+    """User-defined type (UDT).
+
+    .. note:: WARN: Spark Internal Use Only
+    """
+    def __init__(self):
+        raise ContributionsAcceptedError
+
+    @classmethod
+    def typeName(cls) -> str:
+        return cls.__name__.lower()
+
+    @classmethod
+    def sqlType(cls) -> DataType:
+        """
+        Underlying SQL storage type for this UDT.
+        """
+        raise NotImplementedError("UDT must implement sqlType().")
+
+    @classmethod
+    def module(cls) -> str:
+        """
+        The Python module of the UDT.
+        """
+        raise NotImplementedError("UDT must implement module().")
+
+    @classmethod
+    def scalaUDT(cls) -> str:
+        """
+        The class name of the paired Scala UDT (could be '', if there
+        is no corresponding one).
+        """
+        return ""
+
+    def needConversion(self) -> bool:
+        return True
+
+    @classmethod
+    def _cachedSqlType(cls) -> DataType:
+        """
+        Cache the sqlType() into class, because it's heavily used in `toInternal`.
+        """
+        if not hasattr(cls, "_cached_sql_type"):
+            cls._cached_sql_type = cls.sqlType()  # type: ignore[attr-defined]
+        return cls._cached_sql_type  # type: ignore[attr-defined]
+
+    def toInternal(self, obj: Any) -> Any:
+        if obj is not None:
+            return self._cachedSqlType().toInternal(self.serialize(obj))
+
+    def fromInternal(self, obj: Any) -> Any:
+        v = self._cachedSqlType().fromInternal(obj)
+        if v is not None:
+            return self.deserialize(v)
+
+    def serialize(self, obj: Any) -> Any:
+        """
+        Converts a user-type object into a SQL datum.
+        """
+        raise NotImplementedError("UDT must implement toInternal().")
+
+    def deserialize(self, datum: Any) -> Any:
+        """
+        Converts a SQL datum into a user-type object.
+        """
+        raise NotImplementedError("UDT must implement fromInternal().")
+
+    def simpleString(self) -> str:
+        return "udt"
+
+    def __eq__(self, other: Any) -> bool:
+        return type(self) == type(other)
+
+_atomic_types: List[Type[DataType]] = [
+    StringType,
+    BinaryType,
+    BooleanType,
+    DecimalType,
+    FloatType,
+    DoubleType,
+    ByteType,
+    ShortType,
+    IntegerType,
+    LongType,
+    DateType,
+    TimestampType,
+    TimestampNTZType,
+    NullType,
+]
+_all_atomic_types: Dict[str, Type[DataType]] = dict((t.typeName(), t) for t in _atomic_types)
```

Comment:
nitpick: use dictionary comprehensions

Reply:
I appreciate the feedback but this is taken directly from their implementation, I'd rather not touch code that works